### ISSUE TITLE
2.4.0 Loginfile Update Process - Unreliable gamebuild and loginfile versioncheck

### ISSUE THREAD

**Issue by WolfSabian** [TheCheatsrichter/Gw2_Launchbuddy#249]
Currently Launchbuddy is not performing game updates.  I have to launch the game normally and update it, then start launchbuddy.  This then requires me to repair at least 1 loginfile before use.

Anyone else having this issue?

**Comment by Marke3000** [comment]
Im haivng a similar issue , im waiting for a response for my post too

**Comment by Vysil** [comment]
I've had a lot of issues with new releases lately, what I've done is just reverted back to v2.2.0. It works without any issues.

**Comment by maelstromsh** [comment]
I've had a lot of issues with new releases lately, what I've done is just reverted back to v2.2.0. It works without any issues.

**Comment by AZDarkknight** [comment]
Same issue, hasnt updated the last few updates, I choose update on the popup and it just closes, same if I manually choose to update.

**Comment by Marke3000** [comment]
does anyone know how to manually create a login file

**Comment by ScrambledBrain** [comment]
have to confirm more issues with 2.3.9:

I don't the need for ...
> Increased loginfile swap timeout to 20 secs
..., because shouldn't LB find running GW2 (update) processes even each second and wait another second?

However, I begin everytime with a "GW2-64.exe -image" to update GW2 itself 1st.

The I launch LB.

It starts updating loginfile of my 1st account and does it successfully, but the LB popups show me, there is someting wrong with timing.

When it then tries to update loginfile of my 2nd account, it doesn't do it and a popup appears and warns me, that it didn't update and if I proceed, this could be erroneous.
So I pressed no and and the same popup appears again -> endless loop.
So I pressed yes and it skips this account, but says, that this loginfile is not good but also does not update it at later LB starts.
Then I have to "Repair loginfile" manually in LB.

This happens also with my next 2 accounts.

In summary:
loginfile of 1st account : OK
loginfile of 2nd account : FAIL
loginfile of 3rd account : OK
loginfile of 4th account : FAIL

**Comment by WolfSabian** [comment]
Still an issue.

**Comment by anonymous3123** [comment]
Still an issue, can confirm what Scrambled says about every other account. 

1st Good
2nd Fails, needs repair
3rd Good
4th Fails, needs repair
5th Good
6th Fails, needs repair
7th Good
8th Fails, needs repair

**Comment by WolfSabian** [comment]
For me it hasn't been consistent with "every other". I have 4 accounts and on average only 1 has required repair, athough lately it has been 2.

**Comment by akeck** [comment]
> 
> 
> I've had a lot of issues with new releases lately, what I've done is just reverted back to v2.2.0. It works without any issues.

does it work with the current XML file? Or do I need to re-create all accounts again?

**Comment by WolfSabian** [comment]
> Still an issue, can confirm what Scrambled says about every other account.
Seems this is the case for me with the last couple updates.  It wasn't always so, but seems it is now.

**Comment by TheCheatsrichter** [comment]
Fix should be live with 2.4.0.
https://github.com/TheCheatsrichter/Gw2_Launchbuddy/commit/e152d3cced27dfb39cf9c343b5b84a4e7f42f1ac

Please report back if you still have persistent errors.

**Comment by WolfSabian** [comment]
> 
> 
> Fix should be live with 2.4.0.
> [e152d3c](https://github.com/TheCheatsrichter/Gw2_Launchbuddy/commit/e152d3cced27dfb39cf9c343b5b84a4e7f42f1ac)
> 
> Please report back if you still have persistent errors.

Still requires me to update the game outside of LB manually before starting LB.  LaunchBuddy only updates loginfiles.

**Comment by TheCheatsrichter** [comment]
The remaining loginfiles get correctly? (not only every second one)
My guess is that even after a faulty gameclient update the gw2 client seems to update the build version in its xml file.
LB reads that file to gather information and therefore seems to think that your game is allready updated correctly.

To shine some light in the process LB works as follows:

1. Load the gameclient xml file and get the current build version
2. Check the official gw2 API for the newest build version
3. If API version > local version --> update gameclient
3.1 Run the gameclient with "-image" to update the gameclient and automatically close it when its done
3.2 Reimport the game xml file and compare it to the previous version which was gathered from the xml file
4. Compare each account's loginfile build version to the current gameclient version --> update if version is lower

I will further investigate the gw2 gameclient update process and when it writes to its xml file. Maybe some additional indicator for gameclient versionchecking is needed. However I am currently not sure how to achieve this without reading the gameclients memory.

**Comment by TheCheatsrichter** [comment]
Just as i finished writing this comment I found something interesting:

At the moment something seems to be off with the gw2 API:
- Current newest gameclient version : 117585
- Current API build version: 115267

The API currently is not up to date, leads to the following problem:

Lets say your gameclient is on 116000.
You start Launchbuddy it checks the API which returns 115267 --> 116000 > 115267 --> true;  which means Launchbuddy thinks the game to be up to date! 
If you now manually update the game you will update to 117585.
Rerunning Launchbuddy now also results in no gameupdate, however it will work as intended.

Currently my only option would be to force a gameclient update on every startup which is a sketchy fix...
Hopefully Anet fixes this API bug soon. :(

**Comment by ScrambledBrain** [comment]
The issue with the API was already reported at https://github.com/TheCheatsrichter/Gw2_Launchbuddy/issues/237#issuecomment-849933640

Nicodemus also stated, that he wrote a post to Arenanet in https://github.com/TheCheatsrichter/Gw2_Launchbuddy/issues/237#issuecomment-849938519 , which is posted at https://en-forum.guildwars2.com/topic/96243-gw2-client-build-number-stuck-at-115267/ , but still unanswered and still not fixed by Arenanet.

So, please don't rely on API build version anymore!

--

I also realized, that the timing at updating the Accounts files works at my friend's machine, who has a M.2 disk (very fast).

I have a good old HDD disk for some tools like GW2LB and other stuff and only the Game files on my SDD (faster than HDD, but slower than M.2).

I think, that is the reason with such bad sleep timeouts waiting for updated login files at GW2LB update process.

I recommend at least to check, if login files were changed by the update process within last 5 minutes or so.
If not, don't change the version number in Accs.xml

Maybe also try another update run for that account, if user presses "OK" button or skip over if user presses "Skip" button.

**Comment by TheCheatsrichter** [comment]
I am currently creating some concepts to solve this nasty API bug. The quick fix will be the forced update which I likely will push with 2.4.1. However as I said this solution is time consuming and not really a clean way to handle this, at least in my opinion.


Currently the loginfile build number is only updated when:
- the update actually was successfull and something changed in the loginfiles data.
- the users manually accepts that the loginfile is valid anyway.

```
public bool ValidateUpdate(string OldHash)
        {
            Console.WriteLine($"OldHash= {OldHash} NewHash= {MD5HASH}");
            //Should be called after every Update
            if (OldHash == MD5HASH)
            {
                DialogResult dialogResult = MessageBox.Show(($"The loginfile for {AccountManager.GetAccountByID(int.Parse(this.Name)).Nickname} did not change when updating from {Gw2Build} to {EnviromentManager.GwClientVersion}. Proceed anyway?"), "Loginfile Update", MessageBoxButtons.YesNo);
                if (dialogResult == DialogResult.No)
                {
                    Valid = false;
                    return false;
                }
            }
            gw2build = EnviromentManager.GwClientVersion;
            Valid = File.Exists(EnviromentManager.LBLocaldatsPath+Name+".dat");
            return Valid;
        }
```

**Comment by TheCheatsrichter** [comment]
Force gameclient update is now live with https://github.com/TheCheatsrichter/Gw2_Launchbuddy/releases/tag/2.4.1.
This option can be changed in the LB settings tab if needed.

A third option for the loginfile validation process has currently not been added, as the case: loginfile update failed and you want to retry the same file can be achieved by manual validation and pressing loginfile repairing in the acc settings tab.

**Comment by WolfSabian** [comment]
So far working well.  Thanks for the hard work.

**Comment by ScrambledBrain** [comment]
Updates worked for me also today =)
Thanks for your work.

**Comment by TheCheatsrichter** [comment]
https://github.com/TheCheatsrichter/Gw2_Launchbuddy/commit/c422d5b94c27d93ecb7760ca791547db1c6a0096

Released with 2.4.2

**Comment by ScrambledBrain** [comment]
@TheCheatsrichter : thanks for updating.

But I would not hardcode the old build version from API site in a check:
`if(ClientBuild == "115267") { .. }`

This API build version can change at every time and can get old again at every time later on.
This could also happen to the build version from the asset address.

I would read all build versions from all known sites (API, Asset, maybe more sometimes) at any time.

Then make the decision, which version should be returned:
- If no build version could be read, set `forcegameclientupdate = true` and whatever else is needed to force a game update.
- If 1 or more build versions could be read, **return the highest build number** and do the usual tests, if a game and / or login file update is needed.

Greetz.

**Comment by TheCheatsrichter** [comment]
@ScrambledBrain 

True, true, I was busy fixing the versionswitcher which broke a few days ago and totally forgot to implement the cleaner build version check.

https://github.com/TheCheatsrichter/Gw2_Launchbuddy/commit/20247ca4248b744ed2dab54c82d87d70e3111ef8
https://github.com/TheCheatsrichter/Gw2_Launchbuddy/commit/3c99717ea35a80b6cf51f3a7d6ea1f29b4e956c3

These changes already released with 2.4.2 as its release has been delayed to include the versionsiwtcher fix.

**Comment by TheCheatsrichter** [comment]
No new reports since the fix.
Closing.
